[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://githubtocolab.com/CU-Denver-MathStats-OER/Data-Wrangling-and-Visualization/blob/main/08-Tidying-Data-with-tidyr.ipynb)






# <a name="08-title"><font size="6">Module 08: Tidying Data with `tidyr`</font></a>

---

# <a name="tidyr">The `tidyr` Package</a>

---


![tidyr-logo](https://tidyr.tidyverse.org/logo.png)

The goal of the [`tidyr`](https://tidyr.tidyverse.org/) package is to help create **tidy data** that is structured so that:

1. Each variable is in a single column and no variables share a column.
2. Each observation is in a single row.
3. Each data value has its own cell.


The guidelines for tidy data may seem clear and obvious by now.  Why do we need to worry about tidying data, isn't all data tidy? When or why would we ever encounter messy data out there in the "real world". **Unfortunately, most "real" data is messy and tidy.**

- Data is often organized to make *data entry* easy and not typically with *data analysis* in mind.
- For people that do not spent their time working with data, the princples of tidy data are not evident at all.

<br>  



## <a name="load-tidyverse">Loading `tidyverse` Packages</a>

---

The [`tidyverse`](https://www.tidyverse.org/) is a [collection of packages](https://www.tidyverse.org/packages/) by [Hadley Wickham](https://blog.revolutionanalytics.com/2016/09/tidyverse.html) that "share an underlying design philosophy, grammar, and data structures" of tidy data. We can load individual packages within the `tidyverse` one by one, or we can load all packages at once with the command `library(tidyverse)`.

- The `tidyr` package is one of the core `tidyverse` packages.
- The `dplyr` and `ggplot2` packages are also`tidyverse` packages.


In [1]:
library(tidyverse)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.0.4     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


# <a name="tibble">Tibbles</a>

---


A <font color="dodgerblue">**tibble**</font> is a fundamental structure in the `tidyverse` for interacting with most packages and functions. In the words of the developers of the [`tibble`](https://tibble.tidyverse.org/) package:

> "A <font color="dodgerblue">**tibble**</font>, or `tbl_df`, is a modern reimagining of the data.frame, keeping what time has proven to be effective, and throwing out what is not."


## <a name="diff-tib-df">Differences Between a `tibble` and a `data.frame`</a>

---

A tibble is *more or less* a `data.frame`. For example, typically functions, both inside and outside the `tidyverse`, that take a `data.frame` as an input also will take a tibble as an input. However, a tibble does generally have more restrictive behaviors than a `data.frame`.

- Tibbles have an [enhanced `print()`](#enhanced-print) method which makes them easier to use with large datasets.
- Tibbles do not allow for [partial matching of variable names](#part-match).
- [Subsetting a tibble](#subset-tib) always produces a tibble.
- Variables used to create a tibble must (generally) have the [same length](#var-length).



### <a name="enhanced-print">Enhanced Print</a>

---

When viewing a data frame or tibble as output from a code cell in Colab, by default the entire data set is displayed if possible. If the data set is large, then the first 30 rows and last 30 columns are displayed in a scrolling window which may not be very convenient since that is a lot of raw data to look at!

The `print()` function has enhanced capabilities when the input is a tibble compared to a `data.frame`.

- Tibbles only print as many rows and columns as fit on one screen, supplemented by a summary of the remaining rows and columns.
- Tibbles print the data type of each column.
- We can specify the number of rows to print with the option `n`.
- We can increase the column widths with the option `width`.



In [2]:
# storms is a tibble which has class tbl_df
# tibbles are also stored as a data.frame
class(storms)

[1] "tbl_df"     "tbl"        "data.frame"

In [3]:
storms_df <- as.data.frame(storms)  # convert to a data.frame
class(storms_df)  # returns the class attribute as only data.frame

[1] "data.frame"

In [ ]:
# storms_df is a data.frame, not a tibble
# the print() function prints all rows and columns
# too ... much ... information!!!
print(storms_df)

In [ ]:
# in colab this prints first and last 30 rows and
# we see some summary output on top
# still ... too ... much ... information!!!
storms_df

In [6]:
#storms <- as_tibble(storms_df)  # convert to a tibble
class(storms)

[1] "tbl_df"     "tbl"        "data.frame"

In [7]:
# enhanced print for a tibble
#print(storms)
print(storms, n = 5)
#print(storms, n = 5, width = Inf)

# A tibble: 19,537 × 13
  name   year month   day  hour   lat  long status       category  wind pressure
  <chr> <dbl> <dbl> <int> <dbl> <dbl> <dbl> <fct>           <dbl> <int>    <int>
1 Amy    1975     6    27     0  27.5 -79   tropical de…       NA    25     1013
2 Amy    1975     6    27     6  28.5 -79   tropical de…       NA    25     1013
3 Amy    1975     6    27    12  29.5 -79   tropical de…       NA    25     1013
4 Amy    1975     6    27    18  30.5 -79   tropical de…       NA    25     1013
5 Amy    1975     6    28     0  31.5 -78.8 tropical de…       NA    25     1012
# ℹ 19,532 more rows
# ℹ 2 more variables: tropicalstorm_force_diameter <int>,
#   hurricane_force_diameter <int>


### <a name="part-match">Partial Matching of Variable Names</a>

---



In [8]:
# create data.frame
dtf <- data.frame(numbers = 1:4, letters = letters[1:4])
str(dtf)

'data.frame':	4 obs. of  2 variables:
 $ numbers: int  1 2 3 4
 $ letters: chr  "a" "b" "c" "d"


In [9]:
# partial matching of variable name
# will still extract the column
dtf$let

[1] "a" "b" "c" "d"

In [10]:
# create tibble
tbl <- tibble(numbers = 1:4, letters = letters[1:4])
str(tbl)

tibble [4 × 2] (S3: tbl_df/tbl/data.frame)
 $ numbers: int [1:4] 1 2 3 4
 $ letters: chr [1:4] "a" "b" "c" "d"


In [11]:
# partial matching of variable name
tbl$let

Warning message:
“Unknown or uninitialised column: `let`.”


NULL

### <a name="subset-tib">Subset of a Tibble is Always a Tibble</a>

---

Subsetting a `data.frame` can return a vector, which can create edge-cases that results in a bug in your code!




In [12]:
# subset a single column from a data.frame and
# the output is a vector
dtf[ , c("letters")]

[1] "a" "b" "c" "d"

In [13]:
# subset a single row from a data.frame and
# the output is a data.frame
dtf[1, ]

numbers letters
1 1       a

In [14]:
# subset of a data.frame is a vector
dtf[1:2, c("letters")]

[1] "a" "b"

In [15]:
# subset of a data.frame is a data.frame
dtf[1:2, 1:2]

numbers letters
1 1       a      
2 2       b

In [16]:
# subset of a tibble is a tibble
#tbl[, c("letters")]
#tbl[1, ]
#tbl[1:2, c("letters")]
tbl[1:2, 1:2]

numbers letters
1 1       a      
2 2       b

### <a name="var-length">Variable Length</a>

---

- Unlike a `data.frame`, tibbles don't generally "recycle" the values of a vector when creating a data frame.
- Tibbles only recycle vectors of length 1, i.e., it will recycle a vector of length 1 so that the new variable has length matching the other variables.

In [17]:
# add a new column to dtf
dtf$newv <- 1:2
# print dtf
dtf

numbers letters newv
1 1       a       1   
2 2       b       2   
3 3       c       1   
4 4       d       2

In [ ]:
# add a new column to tbl of wrong length (not OK)
tbl$newv <- 1:2

In [20]:
# add a new column by recyclying a single value (OK)
tbl$newv <- 1
# add a new column with the same length as other columns (OK)
#tbl$newv <- rep(1:2, times = 2)
# print tbl
tbl

numbers letters newv
1 1       a       1   
2 2       b       1   
3 3       c       1   
4 4       d       1

# <a name="main-tidy">Tidying Data with `tidyr`</a>

---

Before we can tidy our data, **first we need to identify what are the underlying variables and observations of our data**.

Next, tidying big data sets may require you to do seemingly odd steps such as gathering information from multiple columns into a single column (by adding more rows to the data set) and then spreading information from another column into multiple columns (and reducing the number of rows again).


Here are some tips for tidying a data frame:

1. Identify the ways in which your data are not tidy.
2. Identify a single aspect of the data that you want to make tidy (or at least closer to being tidy).
3. Tidy that single aspect of the data.
4. Repeat steps 2 and 3 until the data are tidy
5. If desired, combine the many tidying steps into a single command using the pipe, `|>`.



## <a name="length">Lengthening Data Frames</a>

---

The `pivot_longer()` function gathers multiple columns into a single column. The resulting data frame will have:

- More rows (longer) and
- Fewer columns.


The following arguments can be entered in the `pivot_longer()` function:


- `data`: the data frame to pivot
- `cols`: the columns from which to gather values
- `names_to`: the column name for the column that the names of the gathered columns are placed
- `values_to`: the name of the column to which the values are gathered




In [21]:
table4a

country     1999   2000  
1 Afghanistan    745   2666
2 Brazil       37737  80488
3 China       212258 213766

### <a name="ex-len">Example: Lengthenging Data</a>

---

- The variables are `year`, `country`, and `cases`.
 - We need to create a column for year.
 - Thus each country will have two rows (one for 1999 and one for 2000).
- The result will have three columns and twice as many rows.


| country | year | cases |
|---------|------|-------|
| Afghanistan | 1999 | 745 |
| Afghanistan | 1999 | 2666 |

<br>  

To tidy `table4a`, we must lengthen it by combining columns `1999` and `2000`. We add a new column named `year` where we indicate from which original column the value is from. We combine the values from the `1999` and `2000` columns into a new column named `cases`.



In [22]:
pivot_longer(table4a,
             c("1999", "2000"),  # combine columns 1999 and 2000
             names_to = "year",  # store variable names to column named year
             values_to = "cases"  # store values to column named cases
)

country     year cases 
1 Afghanistan 1999    745
2 Afghanistan 2000   2666
3 Brazil      1999  37737
4 Brazil      2000  80488
5 China       1999 212258
6 China       2000 213766

In [ ]:
table4a |>
  pivot_longer(
    c("1999", "2000"),  # combine columns 1999 and 2000
    names_to = "year",  # store variable names to column named year
    values_to = "cases"  # store values to column named cases
  )

## <a name="quest1">Question 1</a>

---

Run the code cell below to display the data in `table4b`. Then use the `pivot_longer()` function to tidy the data and display the result to the screen.

<br>  




In [23]:
# we want to combine 1999 and 2000 into one column
# create a column for year
# thus each country will have two rows
table4b

country     1999       2000      
1 Afghanistan   19987071   20595360
2 Brazil       172006362  174504898
3 China       1272915272 1280428583

### <a name="sol1">Solution to Question 1</a>

---

<br>  


### <a name="len-pattern">Lengthening: Data in Variable Names</a>

---

Consider a more complex situation with the larger `who2` data set from which the earlier tables are constructed. The data in `who2` is a "subset of data from the [World Health Organization Global Tuberculosis Report](https://www.who.int/teams/global-tuberculosis-programme/data)", and the data set has 7,240 observations and 60 variables.


The columns `country` and `year` are clear. The other columns are named according to codes from the [World Health Organization](https://www.who.int/teams/global-tuberculosis-programme/data):

> `[method_of_diagnosis]_[gender]_[age_group]`

- The possible `method_of_diagnosis` are: `rel` = relapse, `sn` = negative pulmonary smear, `sp` = positive pulmonary smear, `ep` = extrapulmonary.

- The possible `gender` in the data are: `f` = female, `m` = male.

- The possible `age_group` in the data are: `014` = 0-14 yrs of age, `1524` = 15-24, `2534` = 25-34, `3544` = 35-44 years of age, `4554` = 45-54, `5564` = 55-64, `65` = 65 years or older.

For example, the next column after `year` is `sp_m_014`, which corresponds to positive pulmonary smear for males 0 to 14 years old.

In [ ]:
?who2

In [24]:
print(who2)

# A tibble: 7,240 × 58
   country      year sp_m_014 sp_m_1524 sp_m_2534 sp_m_3544 sp_m_4554 sp_m_5564
   <chr>       <dbl>    <dbl>     <dbl>     <dbl>     <dbl>     <dbl>     <dbl>
 1 Afghanistan  1980       NA        NA        NA        NA        NA        NA
 2 Afghanistan  1981       NA        NA        NA        NA        NA        NA
 3 Afghanistan  1982       NA        NA        NA        NA        NA        NA
 4 Afghanistan  1983       NA        NA        NA        NA        NA        NA
 5 Afghanistan  1984       NA        NA        NA        NA        NA        NA
 6 Afghanistan  1985       NA        NA        NA        NA        NA        NA
 7 Afghanistan  1986       NA        NA        NA        NA        NA        NA
 8 Afghanistan  1987       NA        NA        NA        NA        NA        NA
 9 Afghanistan  1988       NA        NA        NA        NA        NA        NA
10 Afghanistan  1989       NA        NA        NA        NA        NA        NA
# ℹ 7,230 more ro

The variables in a tidy data frame could be:

- `country`
- `year`
- `method`
- `gender`
- `age`
- `cases`

Every column, other than `country` and `year`, we want to split into four columns:

- One column indicates the `method` from the first prefix.
- After the first `_`, we get the class for `gender`.
- After the next `_`, we get the class for `age`.
- The values in the column give the number assigned to the `cases` column.

We can add the `names_sep` argument to `pivot_longer()` to indicate a separator used to indicate each new column.



In [25]:
who2 |>
  pivot_longer(
    cols = !(country:year),  # do not gather country and year
    names_to = c("method", "gender", "age"),  # new column names
    names_sep = "_",  # column separator
    values_to = "cases"  # column name where values stored
  ) |>
  print()

# A tibble: 405,440 × 6
   country      year method gender age   cases
   <chr>       <dbl> <chr>  <chr>  <chr> <dbl>
 1 Afghanistan  1980 sp     m      014      NA
 2 Afghanistan  1980 sp     m      1524     NA
 3 Afghanistan  1980 sp     m      2534     NA
 4 Afghanistan  1980 sp     m      3544     NA
 5 Afghanistan  1980 sp     m      4554     NA
 6 Afghanistan  1980 sp     m      5564     NA
 7 Afghanistan  1980 sp     m      65       NA
 8 Afghanistan  1980 sp     f      014      NA
 9 Afghanistan  1980 sp     f      1524     NA
10 Afghanistan  1980 sp     f      2534     NA
# ℹ 405,430 more rows


## <a name="widen">Widening Data Frames</a>

---

The `pivot_wider()` function spreads a column into multiple columns. The resulting data frame will have:

- More columns (wider) and
- Fewer rows.

The following arguments can be entered in the `pivot_wider()` function:



- `data`: the data frame to pivot
- `names_from`: the column from which the names of the new columns will be taken
- `values_from`: the column from which the values of the new columns will be taken



In [26]:
table2

country     year type       count     
1  Afghanistan 1999 cases             745
2  Afghanistan 1999 population   19987071
3  Afghanistan 2000 cases            2666
4  Afghanistan 2000 population   20595360
5  Brazil      1999 cases           37737
6  Brazil      1999 population  172006362
7  Brazil      2000 cases           80488
8  Brazil      2000 population  174504898
9  China       1999 cases          212258
10 China       1999 population 1272915272
11 China       2000 cases          213766
12 China       2000 population 1280428583

### <a name="ex-wide">Example: Widening Data</a>

---


- The variables are `year`, `country`, `cases`, and `population`.
- The result will have four columns and half as many rows.


To tidy `table2`, we must widen it. We need to take the new column names from the `type` column and the values from the `count` column.


<br>  



In [27]:
pivot_wider(table2,
            names_from = type,  # get new column names from type
            values_from = count  # get values from count column
)

country     year cases  population
1 Afghanistan 1999    745   19987071
2 Afghanistan 2000   2666   20595360
3 Brazil      1999  37737  172006362
4 Brazil      2000  80488  174504898
5 China       1999 212258 1272915272
6 China       2000 213766 1280428583

In [ ]:
table2 |>
  pivot_wider(
    names_from = type,  # get new column names from type
    values_from = count  # get values from count column
  )

# <a name="merge">Merging and Joining Data Frames</a>

---

A common task a data analyst encounters is merging multiple data sets with some overlapping variables. For example, we may download tuberculosis data from the World Health Organization and download separate economic data from the World Bank, and then we may want to study possible associations between variables in both data sets.

To illustrate this task, we consider the two smaller, tidy data frames we constructed earlier with tuberculosis and population data that we assig to  `tbl4a` and `tbl4b`, respectively, in the code cells below.

The most streamlined solutions for this are to use the `base::merge` or `dplyr::*_join`functions to join the columns of the two data frames `by` their shared columns.




In [28]:
tbl4a <- table4a |>
  pivot_longer(
    c("1999", "2000"),  # combine columns 1999 and 2000
    names_to = "year",  # store variable names to column named year
    values_to = "cases"  # store values to column named cases
  )
print(tbl4a)

# A tibble: 6 × 3
  country     year   cases
  <chr>       <chr>  <dbl>
1 Afghanistan 1999     745
2 Afghanistan 2000    2666
3 Brazil      1999   37737
4 Brazil      2000   80488
5 China       1999  212258
6 China       2000  213766


In [29]:
tbl4b <- table4b |>
  pivot_longer(
    c("1999", "2000"),  # combine columns 1999 and 2000
    names_to = "year",  # store variable names to column named year
    values_to = "population"  # store values to column named population
  )
print(tbl4b)

# A tibble: 6 × 3
  country     year  population
  <chr>       <chr>      <dbl>
1 Afghanistan 1999    19987071
2 Afghanistan 2000    20595360
3 Brazil      1999   172006362
4 Brazil      2000   174504898
5 China       1999  1272915272
6 China       2000  1280428583


## <a name="merge-func">`merge()`</a>

---

By default, the `merge` function determines the identical columns in both data frames and joins the rows horizontally (removing redundant information). Alternatively, you can specify columns `by` which to join the rows.

We illustrate usage of the `merge` function below.

- The `x` and `y` arguments are the data frames we want to merge.
- The `by` argument indicates the shared columns we want to match the rows by.
- There are some other fancy arguments you can use if there are matching columns in the data frames that have different names.

Since `merge` is a **base** R function, the result is a `data.frame`.




In [30]:
merge(x = tbl4a, y = tbl4b)

country     year cases  population
1 Afghanistan 1999    745   19987071
2 Afghanistan 2000   2666   20595360
3 Brazil      1999  37737  172006362
4 Brazil      2000  80488  174504898
5 China       1999 212258 1272915272
6 China       2000 213766 1280428583

In [ ]:
merge(x = tbl4a, y = tbl4b, by = c("country", "year"))

## <a name="use_join">`*_join()`</a>

---

If we wish to stay within the `tidyverse` ecosystem, we can use one of the `*_join` functions in `dplyr`. These functions include `x`, `y`, and `by` arguments that work identically to those in the `merge` function.

- `left_join` merges the rows of `y` that match the rows of `x` into `x`.
- `right_join` merges the rows of `x` that match the rows of `y` into `y`.
- `inner_join` merges the rows of `x` and `y` that are in both data frames.
- `full_join` merges the rows of `x` and `y` that are in either data frame.
- See `?inner_join` for more examples.

In this example, since rows of the two data frames match exactly, then any of the joins produce identical results. We will use `full_join`, since anecdotally, it is the most common type of join used.


In [31]:
full_join(tbl4a, tbl4b)

Joining with `by = join_by(country, year)`


country     year cases  population
1 Afghanistan 1999    745   19987071
2 Afghanistan 2000   2666   20595360
3 Brazil      1999  37737  172006362
4 Brazil      2000  80488  174504898
5 China       1999 212258 1272915272
6 China       2000 213766 1280428583

# <a name="separate">Separating Columns</a>

---


The `rate` column of `table3` included both `cases` and `population`. We want to separate these values.



In [32]:
table3

country     year rate             
1 Afghanistan 1999 745/19987071     
2 Afghanistan 2000 2666/20595360    
3 Brazil      1999 37737/172006362  
4 Brazil      2000 80488/174504898  
5 China       1999 212258/1272915272
6 China       2000 213766/1280428583

A column can be separated into multiple columns while retaining the same number of rows using the `separate()` function. Important arguments include:

- `data`: the data frame to manipulate
- `col` the column to separate
- `into`: the names of the variables into which the column will be separated
- `sep`: the character or position by which to separate `col`
- `remove`: a logical value indicating whether `col` will be removed after separation. The default is `TRUE`.



In [33]:
table3 |>
  separate(
    col = rate,  # separate the rate column
    into = c("cases", "population"),  # into two columns with these names
    sep = "/"  # the character by which to separate the values in rate
  )

country     year cases  population
1 Afghanistan 1999 745    19987071  
2 Afghanistan 2000 2666   20595360  
3 Brazil      1999 37737  172006362 
4 Brazil      2000 80488  174504898 
5 China       1999 212258 1272915272
6 China       2000 213766 1280428583

# <a name="unite">Uniting Columns</a>

---


The `unite()` function unites multiple columns of a data frame while retaining the same number of rows. Important arguments include:

- `data`: the data frame to manipulate
- `col` the name of the new column for the united values
- `...`: the columns to unite
- `sep`: the delimiter to separate the values in the columns
- `remove`: a logical value indicating whether the columns in `...` will be removed after unification. The default is `TRUE`.


We use the `unite` function to unite the `century` and `year` columns of `table5`.



In [34]:
# century and year should be united
table5

country     century year rate             
1 Afghanistan 19      99   745/19987071     
2 Afghanistan 20      00   2666/20595360    
3 Brazil      19      99   37737/172006362  
4 Brazil      20      00   80488/174504898  
5 China       19      99   212258/1272915272
6 China       20      00   213766/1280428583

In [35]:
table5 |>
  unite(
    col = "year", "century", "year",  # new_col, old_col1, old_col2, ...
    sep = ""  # do not separate the joined values
  )

country     year rate             
1 Afghanistan 1999 745/19987071     
2 Afghanistan 2000 2666/20595360    
3 Brazil      1999 37737/172006362  
4 Brazil      2000 80488/174504898  
5 China       1999 212258/1272915272
6 China       2000 213766/1280428583

## <a name="quest2">Question 2</a>

---

The output generated by the previous code cell is not tidy. Pipe together commands to represent the data in `table5` as tidy data.

<br>  


### <a name="sol2">Solution to Question 2</a>

---

<br>  


## <a name="CC License">Creative Commons License Information</a>
---

![Creative Commons
License](https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png)

Materials created by the [Department of Mathematical and Statistical Sciences at the University of Colorado Denver](https://github.com/CU-Denver-MathStats-OER/)
and is licensed under a [Creative Commons
Attribution-NonCommercial-ShareAlike 4.0 International
License](http://creativecommons.org/licenses/by-nc-sa/4.0/).